Fix gatt notify multi

laxiLang opened the pull request:
This implements multiple Handle Value Notification (HVN) in one ATT_MULTIPLE_HANDLE_VALUE_NTF PDU. Core Spec 5 .3 | Vol 3, Part F, 3.4.7.4 section describes as this new PDU allows the GATT server to send the values of two or more attributes in a single PDU in the list of tuples of (handle, len, value). However, the Zephyr host currently supports multi HVN but it does not utilize the new ATT PDU of ATT_MULTIPLE_HANDLE_VALUE_NTF=0x23 fully. It is sending only one HVN per PDU, and it is no different from the existing regular single HVN with ATT_HANDLE_VALUE_NTF=0x1B.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/40664

Thalley commented:
This is a copy of https://github.com/zephyrproject-rtos/zephyr/pull/45530

Thalley commented:
@laxiLang Has review comments in the other PR been addressed in this? i.e. is it ready for review?

laxiLang commented:
> @laxiLang Has review comments in the other PR been addressed in this? i.e. is it ready for review?
@Thalley 
yes, it is ready for review.
1. updates the multi notify in GATT
2. add one babbelsim test

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
+			if (!gatt_find_by_uuid(&data, params[i].uuid)) {
+				return -ENOENT;
+			}
+			params[i].attr = data.attr;
```
I know this is part of the existing solution, but I highly dislike updating the parameters like this (but no use fixing it just for notify_mult).

jori-nordic replied:
Yeah I recently encountered it in the gatt subscribed check, it was pretty misleading in gdb. I would be o.k. with either having a commit fixing this in the whole of gatt.c or merging it this way and then making a subsequent PR that fixes this kind of behavior.
I can draft that PR if you don't have the time @laxiLang , just tell me.

Thalley replied:
@jori-nordic Unfortunately I think this is a general issue in GATT. It's quite typical for GATT to modify the data in the parameter structs (for better or worse). 

It will be a significant API change to fix, so at this point the best course of action might be to add additional documentation about it.

laxiLang replied:
@jori-nordic if you help with the proposal for a fix, do appreciate it.

jori-nordic replied:
@Thalley would you be ok merging this as-is (I mean this comment specifically) if I create a doc PR right after ? I don't want to add more noise to that one.

Thalley replied:
Yes, that is fine

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
+				return -ENOENT;
+			}
+			params[i].attr = data.attr;
+		} else {
+			if (!data.handle) {
+				return -ENOENT;
+			}
+		}
```
```suggestion
		} else if (!data.handle) {
			return -ENOENT;
		}
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
-			return -ENOMEM;
+		/* Calculate the ATT_MULTIPLE_HANDLE_VALUE_NTF pdu length */
+		pdu_len += sizeof(*nfy) + params[i].len;
+		if (pdu_len > UINT16_MAX) {
```
This check cannot fail if `pdu_len` is a `uint16_t`. Please change `pdu_len` to `size_t`

alwa-nordic replied:
There is also `u16_add_overflow`.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
-	nfy->len = sys_cpu_to_le16(params->len);
+		/* Check if attribute is a characteristic then adjust the handle */
+		if (!bt_uuid_cmp(params[i].attr->uuid, BT_UUID_GATT_CHRC)) {
+			struct bt_gatt_chrc *chrc = params[i].attr->user_data;
```
```suggestion
			const struct bt_gatt_chrc *chrc = params[i].attr->user_data;
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
+		(void)memcpy(nfy->value, params[i].data, params[i].len);
+	}
+
+	bt_att_set_tx_meta_data(buf, params[0].func, params[0].user_data);
+	return bt_att_send(conn, buf);
```
```suggestion
	bt_att_set_tx_meta_data(buf, params[0].func, params[0].user_data);
	
	return bt_att_send(conn, buf);
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2542,20 +2531,40 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
+	__ASSERT(params->attr || params->uuid, "invalid parameters\n");
+	for (uint16_t i = 1; i < num_params; i++) {
+		__ASSERT(params[0].func == params[i].func, "invalid parameters\n");
+		__ASSERT(memcmp(params[0].user_data, params[i].user_data, 1) == 0,
```
Why are we doing a `memcmp` of the first byte here instead of a pointer comparison?

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2542,20 +2531,40 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
+		return gatt_notify_mult(conn, num_params, params);
+	}
+
+	BT_ERR("Multiple variable length notifications is not supported");
+	return -EOPNOTSUPP;
```
I don't think it is necessary to log an error when we are directly returning an error code. 

```suggestion
	BT_DBG("Multiple variable length notifications is not supported");
	
	return -EOPNOTSUPP;
```

alwa-nordic replied:
In that case, I think we need to document this error code in the documentation of `bt_gatt_notify_multiple`. Anyway, I found the message a bit confusing. I would say "The peer does not have "Multiple Handle
Value Notifications" in its Client Supported Features."

Thalley replied:
Or perhaps simply `BT_DBG("Multiple variable length notifications is not supported by peer");`

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/src/gatt_server_test.c:
```diff
@@ -163,6 +163,40 @@ static inline void long_notify(void)
+	do {
+		err = bt_gatt_notify_multiple(g_conn, ARRAY_SIZE(params), params);
+
+		if (err == -ENOMEM) {
+			k_sleep(K_MSEC(10));
+		} else if (err) {
```
Don't we also get `ENOMEM` if the TX buffer size isn't large enough? In that case we'll just try indefinitely.

laxiLang replied:
@Thalley I see it only returns ENOMEM when calling bt_att_create_pdu failed.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/src/gatt_server_test.c:
```diff
@@ -208,13 +242,63 @@ static void test_main(void)
+	WAIT_FOR_FLAG(flag_short_subscribe);
+	WAIT_FOR_FLAG(flag_long_subscribe);
+
+	for (int i = 0; i < NOTIFICATION_COUNT / 2; i++) {
+		multiple_notify();
+	}
+
+	while (num_notifications_sent < NOTIFICATION_COUNT / 2) {
```
If we assume that `NOTIFICATION_COUNT` is dividable by 2, we should consider adding a BUILD_ASSERT or something to ensure that.

laxiLang replied:
BUILD_ASSERT has been already added

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_notify/test_scripts/multi_notify.sh:
```diff
@@ -0,0 +1,40 @@
+ compile it?)\e[39m"
+    exit 1
+  fi
+  timeout 120 $@ & process_ids="$process_ids $!"
```
`120` is a long timeout for such as simple test. Suggest something shorter like 10 or 20 seconds.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
+		(void)memcpy(nfy->value, params[i].data, params[i].len);
+	}
+
+	bt_att_set_tx_meta_data(buf, params[0].func, params[0].user_data);
```
Indeed. All the individual `params` should have the callback invoked. But where do we store the whole list of `params`? We could..
 - ..add a `_internal.callback_chain` field to `struct bt_gatt_notify_params`. But then we again use the `params` struct to store temporary data, which as commented further up, is not great design.
 - ..require that the `bt_gatt_notify_params` are wrapped in a new struct specialized for multi-notifications.
 - ..somehow manage the storage for it in gatt.c.

Thalley replied:
It is OK for me, as long as either create a TODO or a GH issue that documents the wish to improve this. 

My biggest worry with this API (which pretty much also existing beforehand), is that as an application, actually using the `bt_gatt_notify_multiple` instead of just calling `bt_gatt_notify` multiple time is so much more of a hassle, as well as being more limited that I do not have many intentions of actually using it as is (but at least we'll fix the issues the current implementation has).

jori-nordic replied:
yep agreed, the multiple api is a bit useless right now, I think we'll wait for actual use cases before reworking it, does that sound reasonable ? I'd like to get some feedback on how the users intend on using it, then we can make the necessary adjustments.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2542,20 +2530,40 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
 	__ASSERT(params, "invalid parameters\n");
-	__ASSERT(num_params, "invalid parameters\n");
-	__ASSERT(params->attr, "invalid parameters\n");
+	__ASSERT(num_params >= 2, "invalid parameters\n");
```
The spec allows sending a single notification in a notify_multiple. Should we not allow it?

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2542,20 +2530,40 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
+	__ASSERT(num_params >= 2, "invalid parameters\n");
+	__ASSERT(params->attr || params->uuid, "invalid parameters\n");
+	for (uint16_t i = 1; i < num_params; i++) {
+		__ASSERT(params[0].func == params[i].func, "invalid parameters\n");
+		__ASSERT(memcmp(params[0].user_data, params[i].user_data, 1) == 0,
```
I would rather be in favor of documenting the fact that all callbacks must be the same (for now). And then we could make a new PR to change / add an API to allow different callbacks/user_data.

jori-nordic commented:
Can you check that we don't get too many notification_sent callbacks ? Ie I wanna make sure the callback only gets called once per sent PDU instead of once per attr.I know that's the case right now, but wanna have it in regression.
https://github.com/zephyrproject-rtos/zephyr/blob/9fb216d6043a1205901b566256ac6cf3989eb518/tests/bluetooth/bsim_bt/bsim_test_notify/src/gatt_server_test.c#L282-L284

Vudentz commented:
There is actually a better way to implement this with we could aggregate bt_gatt_notify and then use a delayed_work to dispatch, this simplifies the API and would make the life of application so much easier since they don't need to group the handle+data themselves and have more potential in reducing the buffer usage.

Thalley commented:
> There is actually a better way to implement this with we could aggregate bt_gatt_notify and then use a delayed_work to dispatch, this simplifies the API and would make the life of application so much easier since they don't need to group the handle+data themselves and have more potential in reducing the buffer usage.

That is how the current implementation is designed (although not working correctly). 

The issue with using a `k_work` is that if the caller of the notification function has a lower priority than the system workqueue, then it would never aggregate it (as the notifications would be handled before the next was added). 

Using a `delayed_work` would work, even if we just add like 1ms delay or something I think, and does not suffer from the same issue with priority. 

I do like this approach as an application does not need to worry about whether the remote device actually supports multiple notifications (like this change does), and implement a fallback.

laxiLang committed:
Bluetooth: Host: Fix multiple notify issue

Use a single PDU with the list of tuples of notification handle values,
rather than multiple PDUs.

Signed-off-by: Lang Xie <Lang.Xie@nordicsemi.no>

laxiLang committed:
Bluetooth: Host: Add bsim test for GATT multi notify

Checks that bt_gatt_notify_multiple works as expected.

Signed-off-by: Lang Xie <Lang.Xie@nordicsemi.no>

laxiLang committed:
Bluetooth: Host: Correct for compliance check

Updates needed to pass the compliance check.

Signed-off-by: Lang Xie <Lang.Xie@nordicsemi.no>

Thalley commented:
> Or just try with BlueZ for once? It does support EATT and notify multiple for quite a while, it actually does use a timeout handler (10ms) to group the notifications, actually Im surprised it stop working with Zephyr since Im pretty sure I tested Zephyr with BlueZ extensively since I was developing both in parallel.
> 

I made https://github.com/zephyrproject-rtos/zephyr/pull/45676 to quickly test against other devices, and at least with the current implementation it would not aggregate the notifications, but as mentioned, I think that may be a thread priority issue.

jori-nordic commented:
@Vudentz: I'll try with BlueZ this week: with a 1ms delayed work, and also playing with the calling thread priorities to see if it affects it. 
@Thalley we used https://github.com/zephyrproject-rtos/zephyr/pull/46118#issuecomment-1153657519 at the UPF (thanks btw) and it works pretty well, but we did see that without the current PR it did send normal notifications instead of grouped ones.

Thalley commented:
> @Vudentz: I'll try with BlueZ this week: with a 1ms delayed work, and also playing with the calling thread priorities to see if it affects it. @Thalley we used [#46118 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/46118#issuecomment-1153657519) at the UPF (thanks btw) and it works pretty well, but we did see that without the current PR it did send normal notifications instead of grouped ones.

Did the remote device write to the Client Supported Features before you called it? 

We have this check:
```c
static bool gatt_cf_notify_multi(struct bt_conn *conn)
{
	struct gatt_cf_cfg *cfg;

	cfg = find_cf_cfg(conn);
	if (!cfg) {
		return false;
	}

	return CF_NOTIFY_MULTI(cfg);
}
```

which I enabled for temporary testing with Android (Android does _not_ write to the Client Supported Features (CSF/CF))

jori-nordic commented:
So a quick follow-up: raising the priority doesn't cut it. So something else is happening with the current implementation:
https://github.com/jori-nordic/zephyr/blob/a73c87c2f655b06efcd0e24c292d9bcf1d5129a4/subsys/bluetooth/shell/gatt.c#L956
I still get the normal notifications.

Thalley commented:
> So a quick follow-up: raising the priority doesn't cut it. So something else is happening with the current implementation: https://github.com/jori-nordic/zephyr/blob/a73c87c2f655b06efcd0e24c292d9bcf1d5129a4/subsys/bluetooth/shell/gatt.c#L956 I still get the normal notifications.

If you are getting normal notifications then it's not even reaching the code that has the priority issue. Have you tried disabling the `gatt_cf_notify_multi` check?

jori-nordic commented:
> 

Sorry PEBCAK error, I had forgotten to write to the supported features chrc. Now I do get the multiple notifications PDU.
Then this is with the prio boost:
![image](https://user-images.githubusercontent.com/68428270/173818011-982e2a50-27eb-4c7f-870a-0fdfc7b58347.png)

And without:
![image](https://user-images.githubusercontent.com/68428270/173818077-596021e4-1537-41f5-85cf-6ed4b3d36070.png)

We can definitely see that it does in fact fix the issue, we get a longer PDU when we have a higher prio than the syswq.

Thalley commented:
> > 
> 
> Sorry PEBCAK error, I had forgotten to write to the supported features chrc. Now I do get the multiple notifications PDU. Then this is with the prio boost: ![image](https://user-images.githubusercontent.com/68428270/173818011-982e2a50-27eb-4c7f-870a-0fdfc7b58347.png)
> 
> And without: ![image](https://user-images.githubusercontent.com/68428270/173818077-596021e4-1537-41f5-85cf-6ed4b3d36070.png)
> 
> We can definitely see that it does in fact fix the issue, we get a longer PDU when we have a higher prio than the syswq.

Great, then my theory about it being a priority issue seems to hold water, and is fixable by either changing the priorities, or by using a `delayed_work` (we should also ensure that we never send a notify mult with a single handle, and possibly also reject notify mult with a single handle in the stack as well).

jori-nordic commented:
@Thalley so this is spilling the beans on the API doc that aleksander will be writing but here is how we see it:

if `CONFIG_BT_GATT_NOTIFY_MULTIPLE=y`:
- both `gatt_notify` AND `gatt_notify_multiple` use the queuing mechanism (1ms delayed work) and try to use the multiple notification PDU wherever supported by the peer. (there are some additional constraints for this, like having the same callback/user_data). `gatt_notify` falls back on single notification PDUs when those conditions are not satisfied, `gatt_notify_multiple` returns an error in that case.

if `CONFIG_BT_GATT_NOTIFY_MULTIPLE=n` :
- `gatt_notify_multiple` isn't implemented.
- `gatt_notify` and `gatt_notify_multiple` use single notifications

So all in all, this keeps the 'magic'/automatic behavior that we have now, it just ensures the multi notification PDUs actually contain multiple PDUs. The `gatt_notify_multiple` function should be easy enough, but I will have to rework the actual work item to convert the multiple PDU into a single if there is only one notification when the timeout expires.

There is an alternative, which is having `gatt_notify_multiple` just be a convenience function: this way it automatically downgrades to single notification PDUs when the peer doesn't support it or when `CONFIG_BT_GATT_NOTIFY_MULTIPLE=n` . 
We (@alwa-nordic and I) thought it'd be good to have `gatt_notify_multiple` only output multiple notification PDUs, in case the application developer really wants to use those. But if you (and @Vudentz ) think the alternative behavior is better, just tell us.

Thalley commented:
> @Thalley so this is spilling the beans on the API doc that aleksander will be writing but here is how we see it:
> 
> if `CONFIG_BT_GATT_NOTIFY_MULTIPLE=y`:
> 
>     * both `gatt_notify` AND `gatt_notify_multiple` use the queuing mechanism (1ms delayed work) and try to use the multiple notification PDU wherever supported by the peer. (there are some additional constraints for this, like having the same callback/user_data). `gatt_notify` falls back on single notification PDUs when those conditions are not satisfied, `gatt_notify_multiple` returns an error in that case.

Sounds good
 
> if `CONFIG_BT_GATT_NOTIFY_MULTIPLE=n` :
> 
>     * `gatt_notify_multiple` isn't implemented.
> 
>     * `gatt_notify` and `gatt_notify_multiple` use single notifications

I assume you mean that "`gatt_notify` use single notifications" only

> So all in all, this keeps the 'magic'/automatic behavior that we have now, it just ensures the multi notification PDUs actually contain multiple PDUs. The `gatt_notify_multiple` function should mostly stay as-is, but I will have to rework the actual work item to convert the multiple PDU into a single if there is only one notification when the timeout expires.

So the behavior of `gatt_notify` will be affected by `CONFIG_BT_GATT_NOTIFY_MULTIPLE`? 

> There is an alternative, which is having `gatt_notify_multiple` just be a convenience function: this way it automatically downgrades to single notification PDUs when the peer doesn't support it or when `CONFIG_BT_GATT_NOTIFY_MULTIPLE=n` . We (@alwa-nordic and I) thought it'd be good to have `gatt_notify_multiple` only output multiple notification PDUs, in case the application developer really wants to use those. But if you (and @Vudentz ) think the alternative behavior is better, just tell us.

As a service implementer, I do like the idea of having to use just a single `bt_gatt_notify` function (i.e. either `gatt_notify` or `gatt_notify_multiple`), so that the service can be agnostic w.r.t. whether `CONFIG_BT_GATT_NOTIFY_MULTIPLE` is enabled or not. I am OK with having `gatt_notify_multiple` as long as it falls back to single notifications if the peer device does not support it. 

As mentioned in the old PR for this, I would like to avoid having to do

```c
struct bt_gatt_notify_params params[2];

if (IS_ENABLED(CONFIG_BT_GATT_NOTIFY_MULTIPLE)) {
	err = gatt_notify_multiple(NULL, params, ARRAY_SIZE(params));
	if (err) {
		for (size_t i = 0U; i < ARRAY_SIZE(params); i++) {
			err = gatt_notify_cb(NULL, &param[i]);
			if (err) {
				...
			}
		}
	}
} else {
	for (size_t i = 0U; i < ARRAY_SIZE(params); i++) {
		err = gatt_notify_cb(NULL, &param[i]);
		if (err) {
			...
		}
	}
}
```

jori-nordic commented:
Yeah I meant "gatt_notify use single notifications", good catch.

The problem is that if we implement the auto downgrade for `gatt_notify_multiple` in the stack, the API becomes a bit useless, and imo it'd be better to just remove it.
We were thinking more in the sense that if you as an implementer want to absolutely guarantee your notifications ride on a multiple notification PDU, you can use `gatt_notify_multiple`. For the other 99% of cases, there is `gatt_notify` (that will try to use the multiple notification PDU, but not guaranteed). 

So `gatt_notify_multiple` would be more of a "I really know what I want and what I'm doing" API, not a "oh it's convenient, I don't have to write a for loop around `gatt_notify`" kind of API, I don't know if I'm being clear :smile:

jori-nordic commented:
https://github.com/zephyrproject-rtos/zephyr/pull/46118#issuecomment-1157619478 : Then it seems we agree. Let's keep the notify_multi api, I will just add scary warnings in the doc. I will also fix and document the normal notify API to make its behavior clear to the user.
